### ISSUE TITLE
Shut down task executor beans first during factory destruction [SPR-9512]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-framework#14146]
**[Archie Cobbs](https://jira.spring.io/secure/ViewProfile.jspa?name=archie172)** opened **[SPR-9512](https://jira.spring.io/browse/SPR-9512?redirect=false)** and commented

I have a bean factory with this:

```
    <task:annotation-driven executor="taskExecutor" scheduler="taskScheduler"/>

    <task:executor id="taskExecutor" pool-size="2-10" queue-capacity="100" keep-alive="10"/>
    <task:scheduler id="taskScheduler" pool-size="1"/>
```

and I'm occasionally seeing this exception on bean factory shutdown:

```
org.springframework.beans.factory.BeanCreationNotAllowedException: Error creating bean with name 'transactionManager':
    Singleton bean creation not allowed while the singletons of this factory are in destruction (Do not request a bean
    from a BeanFactory in a destroy method implementation!)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:212)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:291)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.determineTransactionManager(TransactionAspectSupport.java:248)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:272)
        at org.springframework.transaction.aspectj.AbstractTransactionAspect.ajc$before$org_springframework_transaction_aspectj_AbstractTransactionAspect$1$2a73e96c(AbstractTransactionAspect.aj:63)
        at com.foobar.DefaultQueueProcessor.runQueue(DefaultQueueProcessor.java:50)
        at sun.reflect.GeneratedMethodAccessor59.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:64)
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:53)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
        at java.util.concurrent.FutureTask$Sync.innerRunAndReset(FutureTask.java:317)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:150)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$101(ScheduledThreadPoolExecutor.java:98)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.runPeriodic(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:204)
        at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
        at java.lang.Thread.run(Thread.java:680)
```

A scheduled task is starting and then dying because the bean factory is shutting down.

It seems like an improvement would be: before shutting down a bean factory, find all beans that perform scheduled tasks and stop them, so they don't generate new activity during bean factory destruction.

---

**Affects:** 3.1.1

9 votes, 12 watchers

**Comment by OleksiiBondar** [comment]
Is there any update on this?

**Comment by limhs25** [comment]
Not Yet? or Newer? Any body here?

**Comment by Sineaggi** [comment]
We're seeing this too. Any update?

**Comment by rustyx** [comment]
Same here with Spring Boot `2.2.4.RELEASE` (Spring `5.2.3.RELEASE`).

From the looks of it, if a new web request comes in at the exact moment the server is shutting down, the dispatcher tries to re-create the already destroyed controller.

```
org.springframework.beans.factory.BeanCreationNotAllowedException: Error creating bean with name 'mainController': Singleton bean creation not allowed while singletons of this factory are in destruction (Do not request a bean from a BeanFactory in a destroy method implementation!)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:208)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.web.method.HandlerMethod.createWithResolvedBean(HandlerMethod.java:336)
        at org.springframework.web.servlet.handler.AbstractHandlerMethodMapping.getHandlerInternal(AbstractHandlerMethodMapping.java:368)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping.getHandlerInternal(RequestMappingHandlerMapping.java:449)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping.getHandlerInternal(RequestMappingHandlerMapping.java:67)
        at org.springframework.web.servlet.handler.AbstractHandlerMapping.getHandler(AbstractHandlerMapping.java:395)
        at org.springframework.web.servlet.DispatcherServlet.getHandler(DispatcherServlet.java:1234)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1016)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:747)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.ajp.AjpProcessor.service(AjpProcessor.java:399)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:860)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1598)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by acsipak** [comment]
Any suggested workarounds?

**Comment by stefanocke** [comment]
As far as I see, there are at least two issues here in the comments:

1. BeanCreationNotAllowedException in scheduled tasks
2. BeanCreationNotAllowedException in web requests

**Regarding scheduled tasks:**

I also experienced BeanCreationNotAllowedExceptions but sometimes also "EntityManager is closed", if the task did some JPA work.
When I lookes deeper into this, I recognized the main cause was, that my task was not **interruptible**. 

So, the following happened:
1. TaskScheduler is stopped and tries to interrupt its scheduled tasks
2. Tasks continue to run, since they are not interruptible.
3. Spring shuts down other beans, like EntityManagerFactory
4. Task try to get beans / resources, that no longer exist -> Exceptions are thrown.

By "not interruptible" I mean, they are not running any oprations mentionend in java.lang.Thread.interrupt().
This is for instance even true if the task does lots of DB work and uses Postgres JDBC driver, since it is pre-NIO. (This explanation is probably not perfect, but I am quite sure my long running DB tasks were definitely not interruptible...)

Thus, a possible solution for scheduled tasks is to make them interruptible b checking for `Thread.current().isInterrupted()` frequently. See for instance https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#2.2

**Additionally**, you should set `spring.task.execution.shutdown.await-termination-period` and `spring.task.execution.shutdown.await-termination-period`. So you compensate for the milliseconds or seconds you scheduled task needs until it checks for `Thread.current().isInterrupted()`  the next time.


**Regarding web requests:**

I have not too much experince with this case but hope that https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.3-Release-Notes#graceful-shutdown would help.

**Comment by stefanocke** [comment]
Some more debugging...  I looked at the disposable bean names and their order by debugging into `destroySingletons()` .
For better  overview I only show some relevant beans here:

```
[... dataSource,
 ..., entityManagerFactory, 
 ..., taskScheduler,
 ... , applicationTaskExecutor, 
 ..., refreshEventListener, 
..., springApplicationAdminRegistrar, ... ]
```

The destruction oder is from last to first element.
As you can see, dataSource and entityManagerFactory are already in good positions. If my tasks depend on them, I can rely on them being destroyed after taskScheduler and applicationTaskExecutor.

However, I got my exception for the bean "refreshEventListener" and also for "springApplicationAdminRegistrar".
As the stacktrace in my previous comment shows, they are just requested because the EventMulticaster ask for all potential EventListener beans. (AFAIU. Excuse me, if incorrect...)
So, my tasks have a very "indirect" dependency on those beans, just because I publish events.
I tend to belive it is not possible to completely detect all those kinds of dependencies.
So maybe, the original title of this issue from 2012 is still valid: "Shut down task executor beans first during factory destruction"

**Comment by jhoeller** [comment]
Superseded by #27090 now, being released today in 6.1 M2.

We're effectively doing what's shown above, triggering an early shutdown on `ContextClosedEvent` for both `ThreadPoolTaskExecutor` and `ThreadPoolTaskScheduler`, as well as for `ScheduledAnnotationBeanPostProcessor`.
